### PR TITLE
Binary search the entries when all suffixes have the same length in a leaf block.

### PR DESCRIPTION
**Pull request by vsop-479**
In scanToTerm phase, binary search may have a better performance when all suffixes have the same length in a leaf block.

### CODE REVIEW

File: lucene/core/src/java/org/apache/lucene/codecs/lucene90/blocktree/SegmentTermsEnumFrame.java [hunk 1]
```diff
@@ -646,6 +648,84 @@ public SeekStatus scanToTermLeaf(BytesRef target, boolean exactOnly) throws IOEx
...
+
+    suffix = suffixLengthsReader.readVInt();
+    int start = nextEnt;
+    int end = entCount - 1;
+    //Binary search the entries (terms) in this leaf block:
+    while (start <= end) {
+      int mid = (start + end) / 2;
+      nextEnt = mid + 1;
+      startBytePos = mid * suffix;
+      // Loop over bytes in the suffix, comparing to the target
```
Comment (jpountz) [inline lucene/core/src/java/org/apache/lucene/codecs/lucene90/blocktree/SegmentTermsEnumFrame.java:681]: Maybe update the comment, it's no longer a loop?

**Review by jpountz — COMMENTED** [review]
Nice, have you been able to measure a speedup with this change?

I'm not sure how well covered same-length terms are covered in our tests, maybe add a new one to `BasePostingsFormatTestCase`? It looks like comments were copy-pasted and need updating too, e.g. references to "scanning" since we're no longer scanning a block in this case.

**Comment by vsop-479** [comment]
@jpountz 
Thanks for your review.
I did a simple performance test, which indexed 1M random UUID's substring(2, 8), got 10 segments, and picked up 1K terms to search. Average Result of 4 times tests: 

Method took:
baseline(scanToTermLeaf)ns | candidate(binarySearchTermLeaf)ns | speedup
-- | -- |--
5,757,121.5 | 4,761,325.5 | 20.9%

Whole search took:
baseline(scan)ns | candidate(binarySearch)ns | speedup
-- | -- |--
162,668,448 | 157,990,611 | 2.9%

In my test case, scanToTerm only took 3.5% of the whole search execute time, so it could only got a small speedup.
I may add this test case to BasePostingsFormatTestCase, or do you have any other idea on test?
I willl update the comment, please have a review.

**Comment by jpountz** [comment]
> I may add this test case to BasePostingsFormatTestCase, or do you have any other idea on test?

1M documents is too much for a unit test, I was thinking of a smaller dataset, e.g. 200 fixed-size IDs and we'd make sure that the binary search works as expected for both `seekCeil` and `seekExact` for every of these 200 terms as well as other terms that compare less than all terms from the dict, greater than all terms of the dict, or are between two terms that exist in the dict?

**Comment by vsop-479** [comment]
@jpountz 
I added a test to BasePostingsFormatTestCase. Please have a review.

File: lucene/test-framework/src/java/org/apache/lucene/tests/index/BasePostingsFormatTestCase.java [hunk 2]
```diff
@@ -367,6 +367,49 @@ public void testGhosts() throws Exception {
...
+    }
+    iw.commit();
+    iw.forceMerge(1);
+
+    DirectoryReader reader = DirectoryReader.open(iw);
+    TermsEnum termsEnum = getOnlyLeafReader(reader).terms("id").iterator();
+    // test seekExact
+    for (int i = 100000; i <= 100400; i++){
+      if (i % 2 == 1) {
+        assertTrue(termsEnum.seekExact(new BytesRef(i + "")));
```
Comment (jpountz) [inline lucene/test-framework/src/java/org/apache/lucene/tests/index/BasePostingsFormatTestCase.java:394]: maybe assert that the value of `termsEnum.term()` is correct after this call

File: lucene/test-framework/src/java/org/apache/lucene/tests/index/BasePostingsFormatTestCase.java [hunk 3]
```diff
@@ -367,6 +367,49 @@ public void testGhosts() throws Exception {
...
+      } else {
+        assertFalse(termsEnum.seekExact(new BytesRef(i + "")));
+      }
+    }
+    // test seekCeil
+    for (int i = 100000; i < 100400; i++){
+      if (i % 2 == 1) {
+        assertEquals(SeekStatus.FOUND, termsEnum.seekCeil(new BytesRef(i + "")));
```
Comment (jpountz) [inline lucene/test-framework/src/java/org/apache/lucene/tests/index/BasePostingsFormatTestCase.java:402]: maybe assert that the value of `termsEnum.term()` is correct after this call

**Review by jpountz — APPROVED** [review]
I left a minor suggestion but it looks good to me. Can you add a CHANGES entry under 9.5?

**Comment by vsop-479** [comment]
@jpountz Thranks for your review and suggestion. I have added a CHANGES entry and assert term value code. 
Please have a review.

File: lucene/test-framework/src/java/org/apache/lucene/tests/index/BasePostingsFormatTestCase.java [hunk 4]
```diff
@@ -367,6 +367,53 @@ public void testGhosts() throws Exception {
...
+      }
+    }
+    // test seekCeil
+    for (int i = 100000; i < 100400; i++){
+      BytesRef target = new BytesRef(i + "");
+      if (i % 2 == 1) {
+        assertEquals(SeekStatus.FOUND, termsEnum.seekCeil(target));
+        assertEquals(termsEnum.term(), target);
+      } else {
+        assertEquals(SeekStatus.NOT_FOUND, termsEnum.seekCeil(target));
```
Comment (jpountz) [inline lucene/test-framework/src/java/org/apache/lucene/tests/index/BasePostingsFormatTestCase.java:408]: maybe also assert here that the terms enum is positioned on the next term?

**Review by jpountz — APPROVED** [review]
Given the test failure, it looks like you need to run `gradlew tidy` to fix indentation.

**Comment by vsop-479** [comment]
@jpountz Thranks for your commit.
Yes, there seems to be some problem on deleteDocuments. I am trying to figure it out.

**Comment by vsop-479** [comment]
@jpountz I have passed the tests locally, please have a review.

**Comment by jpountz** [comment]
@mikemccand You might want to have a look at this change since (I think) you are one of the most familiar ones with the original code.

**Comment by jpountz** [comment]
I had to revert this change because of test failures, e.g. this seed reproduces on the main branch:

```
gradlew test --tests TestNumericDocValuesUpdates.testSortedIndex -Dtests.seed=4C6E977E1F29E069 -Dtests.locale=khq -Dtests.timezone=Asia/Hong_Kong -Dtests.asserts=true -Dtests.file.encoding=UTF-8
```

It seems like this test exercises some logic that `BasePostingsFormatTestCase` doesn't but I haven't figured out what yet.

**Comment by vsop-479** [comment]
@jpountz I have fixed the test failures. 
But I cant find the new commit in this PR (since it has been merged and reverted?), do I need to open an new PR or some what to show the new commit?

**Comment by jpountz** [comment]
A new PR would be great, thank you for looking into these failures!

**Pull request merged**